src/v/storage/disk_log_impl.cc

```diff
@@ -480,15 +480,26 @@ segment_set disk_log_impl::find_sliding_range(
     // Collect all segments that have stable data.
     segment_set::underlying_t buf;
     for (const auto& seg : _segs) {
+        if (seg->has_appender() || !seg->has_compactible_offsets(cfg)) {
+            // Stop once we get to an unstable segment.
+            break;
+        }
         if (
           new_start_offset
           && seg->offsets().base_offset < new_start_offset.value()) {
             // Skip over segments that are being truncated.
             continue;
         }
-        if (seg->has_appender() || !seg->has_compactible_offsets(cfg)) {
-            // Stop once we get to an unstable segment.
-            break;
+        if (buf.empty() && !seg->has_compactible_data_records()) {
+            // To find the start of the sliding range, skip over segments that
+            // have been effectively entirely compacted away, i.e. who no
+            // longer have data or have only their last batch remaining.
+            //
+            // NOTE: We don't do this past the beginning of the sliding range
+            // so that the rest of the segments are contiguous.
+            seg->mark_as_finished_self_compaction();
+            seg->mark_as_finished_windowed_compaction();
+            continue;
         }
         buf.emplace_back(seg);
     }
```

Comment:
nit: i find it a bit hard to follow these single loops, 
i would prefer something like

```cpp
auto first_non_truncated =new_start_offset? std::ranges::find_if(_segs, [](auto&seg){return seg->offsets().base_offset < new_start_offset.value;}) : _segs.begin();
auto first_compactible=std::ranges::find_if(first_non_truncated, _segs.end(), &segment::had_comactible_data_reconds);
auto end_compactible=std::find_if(first_compactible, _segs.end(), [](auto&seg){return seg->has_appender() || seg->has_comactible_offsets(cfg));

std::for_each(first_non_truncated, first_comactible, [](auto seg&) {seg->mark_as_finished_selg_compaction(); seg->mark_as_finished_windowes_compaction})
auto bug=underlying_t{first_compactible, end_compactible);
```

but it's multi-pass and not really shorter, so it's a matter of preference

Reply:
I agree it's a little tricky, but i'm really not a fan of iterator-based logic. It feels clunky and difficult to read (at least harder to grok than a simple loop IMO)